Mathlib/RingTheory/Flat.lean

```diff
@@ -110,6 +113,22 @@ lemma iff_rTensor_injective :
     rw [← aux]
     simp [h₁ hI]
 
+/-- An `R`-module `M` is flat iff for all ideals `I` of `R`, the tensor product of the
+inclusion `I → R` and the identity `M → M` is injective. See `iff_rTensor_injective` to
+restrict to finitely generated ideals `I`. --/
+theorem iff_rTensor_injective' :
+    Flat R M ↔ (∀ (I : Ideal R), Injective (rTensor M I.subtype)) := by
+  rewrite [Flat.iff_rTensor_injective]
+  refine ⟨fun h I => ?_, fun h I _ => h I⟩
+  letI : AddCommGroup (I ⊗[R] M) := inferInstance -- Type class reminder
+  rewrite [injective_iff_map_eq_zero]
+  intro x hx₀
+  have ⟨J, hfg, hle, y, hy⟩ := exists_fg_le_eq_rTensor_inclusion x
+  apply (fun hy₀ => by rw [hy, hy₀, _root_.map_zero] : y = 0 → x = 0)
+  rewrite [hy, ← comp_apply, ← rTensor_comp] at hx₀
+  letI : AddCommGroup (J ⊗[R] M) := inferInstance -- Type class reminder
+  exact (injective_iff_map_eq_zero _).mp (h hfg) y hx₀
```

Comment:
slightly golfed: removing some redundant parentheses, and use `obtain rfl`, `rTensor_comp_apply`. `rw` is much more idiomatic in mathlib than `rewrite`, in fact I don't recall ever seeing the latter in my life!
```suggestion
    Flat R M ↔ ∀ I : Ideal R, Injective (rTensor M I.subtype) := by
  rw [Flat.iff_rTensor_injective]
  refine ⟨fun h I => ?_, fun h I _ => h I⟩
  letI : AddCommGroup (I ⊗[R] M) := inferInstance -- Type class reminder
  rw [injective_iff_map_eq_zero]
  intro x hx₀
  obtain ⟨J, hfg, hle, y, rfl⟩ := exists_fg_le_eq_rTensor_inclusion x
  rw [← rTensor_comp_apply] at hx₀
  letI : AddCommGroup (J ⊗[R] M) := inferInstance -- Type class reminder
  rw [(injective_iff_map_eq_zero _).mp (h hfg) y hx₀, _root_.map_zero]
```

Reply:
Thanks!

`rewrite` and `rw` are not the same thing (see the `rw` docstring). I prefer to use the one I mean in each case. It's [not a popular position](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/tactic.20docstrings/near/393803105), but it's correct, and I don't want to do it the wrong way just because everyone else does. Not a hill I'll choose to die on, ultimately, but do you mind if apply the rest of the suggestion but change the non-finishing `rw`'s back to `rewrite`s?

Reply:
> do you mind if apply the rest of the suggestion but change the non-finishing `rw`'s back to `rewrite`s?

Done. Let me know if you feel strongly about `rw` vs `rewrite`.

Reply:
Sounds good!